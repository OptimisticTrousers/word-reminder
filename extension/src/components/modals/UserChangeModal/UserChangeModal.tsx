import { CODE_MAX_BYTES, User } from "common";
import CSSModules from "react-css-modules";
import { useNavigate, useOutletContext } from "react-router-dom";

import { ModalContainer } from "../ModalContainer";
import styles from "./UserChangeModal.module.css";

interface Props {
  toggleModal: () => void;
  path: string;
}

export const UserChangeModal = CSSModules(
  function ({ toggleModal, path }: Props) {
    const { user }: { user: User } = useOutletContext();
    const navigate = useNavigate();

    function handleConfirm(formData: FormData) {
      const code = formData.get("code") as string;
      navigate(`/users/update/${path}/${code}`);
      toggleModal();
    }

    function handleCancel() {
      toggleModal();
    }

    return (
      <ModalContainer title="Confirm your email" toggleModal={toggleModal}>
        <form styleName="modal" action={handleConfirm}>
          <p styleName="modal__message">
            Please enter the confirmation code that was sent to{" "}
            <span styleName="modal__bold">{user.email}</span> within 5 minutes.
          </p>
          <label styleName="modal__label">
            Code
            <input
              styleName="modal__input"
              name="code"
              type="text"
              required={true}
              maxLength={CODE_MAX_BYTES}
            />
          </label>
          <div styleName="modal__buttons">
            <button
              styleName="modal__button modal__button--cancel"
              onClick={handleCancel}
            >
              Cancel
            </button>
            <button
              styleName="modal__button modal__button--confirm"
              type="submit"
            >
              Enter Code
            </button>
          </div>
        </form>
      </ModalContainer>
    );
  },
  styles,
  { allowMultiple: true, handleNotFoundStyleName: "log" }
);
