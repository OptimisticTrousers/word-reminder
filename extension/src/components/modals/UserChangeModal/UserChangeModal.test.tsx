import { createRoutesStub, Outlet, useParams } from "react-router-dom";
import { render, screen } from "@testing-library/react";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import userEvent from "@testing-library/user-event";

import { UserChangeModal } from "./UserChangeModal";

vi.mock("../ModalContainer/ModalContainer");

describe("EmailChangeModal", () => {
  const testUser = {
    email: "bob@email.com",
    id: "1",
  };

  const mockToggleModal = vi.fn();

  function setup(path: string) {
    const queryClient = new QueryClient();
    const Stub = createRoutesStub([
      {
        path: "/",
        Component: function () {
          return <Outlet context={{ user: testUser }} />;
        },
        children: [
          {
            path: "/users/update",
            Component: function () {
              return (
                <QueryClientProvider client={queryClient}>
                  <UserChangeModal path={path} toggleModal={mockToggleModal} />
                </QueryClientProvider>
              );
            },
          },
          {
            path: "/users/update/email/:code",
            Component: function () {
              const { code } = useParams();

              return (
                <div>You can change your email thanks to this code: {code}</div>
              );
            },
          },
          {
            path: "/users/update/password/:code",
            Component: function () {
              const { code } = useParams();

              return (
                <div>
                  You can change your password thanks to this code: {code}
                </div>
              );
            },
          },
        ],
      },
    ]);
    return {
      user: userEvent.setup(),
      ...render(<Stub initialEntries={["/users/update"]} />),
    };
  }

  beforeEach(() => {
    vi.clearAllMocks();
  });

  describe("updateEmail", () => {
    const path = "email";

    it("shows form", async () => {
      const { asFragment } = setup(path);

      const message = screen.getByText((_, element) => {
        return (
          element?.textContent ===
          `Please enter the confirmation code that was sent to ${testUser.email} within 5 minutes.`
        );
      });
      expect(message).toBeInTheDocument();
      expect(asFragment()).toMatchSnapshot();
    });

    it("calls the functions to navigate to update email page with token param", async () => {
      const { user } = setup(path);

      const code = "code";
      const codeInput = screen.getByLabelText("Code", { selector: "input" });
      await user.type(codeInput, code);
      const enterCodeButton = screen.getByRole("button", {
        name: "Enter Code",
      });
      await user.click(enterCodeButton);

      const updateCode = await screen.findByText(
        `You can change your email thanks to this code: ${code}`
      );
      expect(updateCode).toBeInTheDocument();
      expect(mockToggleModal).toHaveBeenCalledTimes(1);
      expect(mockToggleModal).toHaveBeenCalledWith();
    });

    it("does not allow the user to submit when the code field is empty", async () => {
      const { user } = setup(path);

      const enterCodeButton = screen.getByRole("button", {
        name: "Enter Code",
      });
      await user.click(enterCodeButton);

      expect(mockToggleModal).not.toHaveBeenCalled();
    });

    it("works when user closes the modal using cancel button", async () => {
      const { user } = setup("updatePassword");

      const cancelButton = screen.getByRole("button", { name: "Cancel" });
      await user.click(cancelButton);

      expect(mockToggleModal).toHaveBeenCalledTimes(1);
      expect(mockToggleModal).toHaveBeenCalledWith();
    });

    it("works when user closes the modal using x button", async () => {
      const { user } = setup(path);

      const modalContainerCloseButton = screen.getByRole("button", {
        name: "Cancel",
      });
      await user.click(modalContainerCloseButton);

      expect(mockToggleModal).toHaveBeenCalledTimes(1);
      expect(mockToggleModal).toHaveBeenCalledWith();
    });
  });

  describe("updatePassword", () => {
    const path = "password";

    it("shows form", async () => {
      const { asFragment } = setup(path);

      const message = screen.getByText((_, element) => {
        return (
          element?.textContent ===
          `Please enter the confirmation code that was sent to ${testUser.email} within 5 minutes.`
        );
      });
      expect(message).toBeInTheDocument();
      expect(asFragment()).toMatchSnapshot();
    });

    it("calls the functions to navigate to update password page with token param", async () => {
      const { user } = setup(path);

      const code = "code";
      const codeInput = screen.getByLabelText("Code", { selector: "input" });
      await user.type(codeInput, code);
      const enterCodeButton = screen.getByRole("button", {
        name: "Enter Code",
      });
      await user.click(enterCodeButton);

      const updateCode = await screen.findByText(
        `You can change your password thanks to this code: ${code}`
      );
      expect(updateCode).toBeInTheDocument();
      expect(mockToggleModal).toHaveBeenCalledTimes(1);
      expect(mockToggleModal).toHaveBeenCalledWith();
    });

    it("does not allow the user to submit when the code field is empty", async () => {
      const { user } = setup(path);

      const enterCodeButton = screen.getByRole("button", {
        name: "Enter Code",
      });
      await user.click(enterCodeButton);

      expect(mockToggleModal).not.toHaveBeenCalled();
    });

    it("works when user closes the modal using cancel button", async () => {
      const { user } = setup("updatePassword");

      const cancelButton = screen.getByRole("button", { name: "Cancel" });
      await user.click(cancelButton);

      expect(mockToggleModal).toHaveBeenCalledTimes(1);
      expect(mockToggleModal).toHaveBeenCalledWith();
    });

    it("works when user closes the modal using x button", async () => {
      const { user } = setup(path);

      const modalContainerCloseButton = screen.getByRole("button", {
        name: "Cancel",
      });
      await user.click(modalContainerCloseButton);

      expect(mockToggleModal).toHaveBeenCalledTimes(1);
      expect(mockToggleModal).toHaveBeenCalledWith();
    });
  });
});
