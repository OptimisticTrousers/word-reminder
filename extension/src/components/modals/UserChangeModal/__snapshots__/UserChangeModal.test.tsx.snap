// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`EmailChangeModal > updateEmail > shows form 1`] = `
<DocumentFragment>
  <div>
    <h2>
      Confirm your email
    </h2>
    <button
      aria-label="Close modal"
    >
      X
    </button>
    <div>
      <form
        action="javascript:throw new Error('A React form was unexpectedly submitted. If you called form.submit() manually, consider using form.requestSubmit() instead. If you\\'re trying to use event.stopPropagation() in a submit event handler, consider also calling event.preventDefault().')"
        class="_modal_ff4fcd"
      >
        <p
          class="_modal__message_ff4fcd"
        >
          Please enter the confirmation code that was sent to 
          <span
            class="_modal__bold_ff4fcd"
          >
            bob@email.com
          </span>
           within 5 minutes.
        </p>
        <label
          class="_modal__label_ff4fcd"
        >
          Code
          <input
            class="_modal__input_ff4fcd"
            maxlength="16"
            name="code"
            required=""
            type="text"
          />
        </label>
        <div
          class="_modal__buttons_ff4fcd"
        >
          <button
            class="_modal__button_ff4fcd _modal__button--cancel_ff4fcd"
          >
            Cancel
          </button>
          <button
            class="_modal__button_ff4fcd _modal__button--confirm_ff4fcd"
            type="submit"
          >
            Enter Code
          </button>
        </div>
      </form>
    </div>
  </div>
</DocumentFragment>
`;

exports[`EmailChangeModal > updatePassword > shows form 1`] = `
<DocumentFragment>
  <div>
    <h2>
      Confirm your email
    </h2>
    <button
      aria-label="Close modal"
    >
      X
    </button>
    <div>
      <form
        action="javascript:throw new Error('A React form was unexpectedly submitted. If you called form.submit() manually, consider using form.requestSubmit() instead. If you\\'re trying to use event.stopPropagation() in a submit event handler, consider also calling event.preventDefault().')"
        class="_modal_ff4fcd"
      >
        <p
          class="_modal__message_ff4fcd"
        >
          Please enter the confirmation code that was sent to 
          <span
            class="_modal__bold_ff4fcd"
          >
            bob@email.com
          </span>
           within 5 minutes.
        </p>
        <label
          class="_modal__label_ff4fcd"
        >
          Code
          <input
            class="_modal__input_ff4fcd"
            maxlength="16"
            name="code"
            required=""
            type="text"
          />
        </label>
        <div
          class="_modal__buttons_ff4fcd"
        >
          <button
            class="_modal__button_ff4fcd _modal__button--cancel_ff4fcd"
          >
            Cancel
          </button>
          <button
            class="_modal__button_ff4fcd _modal__button--confirm_ff4fcd"
            type="submit"
          >
            Enter Code
          </button>
        </div>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
